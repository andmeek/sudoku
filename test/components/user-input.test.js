import { createLocalVue, shallowMount } from '@vue/test-utils'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import UserInput from '../../src/components/user-input.vue'

const localVue = createLocalVue()
localVue.component('font-awesome-icon', FontAwesomeIcon)

describe('UserInput.vue', () => {
  let wrapper, board

  beforeEach(() => {
    board = genTestBoard()
    wrapper = shallowMount(UserInput, {
      localVue,
      propsData: {
        board: board
      }
    })
  })

  test('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('toggles the selection if hit again', () => {
    expect(wrapper.find('.selected').exists()).toBeFalsy()

    wrapper.find('[value="1"]').trigger('click')
    expect(wrapper.find('.selected').attributes('value')).toEqual('1')

    wrapper.find('[value="1"]').trigger('click')
    expect(wrapper.find('.selected').exists()).toBeFalsy()
  })

  test('disables completed numbers', () => {
    board.tiles.forEach((tile) => {
      if(tile.actualValue == 1)
        tile.userValue = tile.actualValue
    })

    wrapper.vm.$forceUpdate()
    expect(board.numberCompleted('1')).toBeTruthy()
    expect(wrapper.find('.disabled').attributes('value')).toEqual('1')
    expect(wrapper.find('.disabled').attributes('disabled')).toEqual('disabled')

    expect(wrapper.html()).toMatchSnapshot()

    expect(wrapper.vm.current).toBeNull()
  })

  test('unselects completed numbers if selected', () => {
    wrapper.find('[value="1"]').trigger('click')
    expect(wrapper.find('.selected').exists()).toBeTruthy()

    board.tiles.forEach((tile) => {
      if(tile.actualValue == 1)
        tile.userValue = tile.actualValue
    })

    wrapper.vm.$forceUpdate()
    expect(wrapper.find('.selected').exists()).toBeFalsy()
    expect(wrapper.emitted().inputchanged).toBeTruthy()
  })

  test('highlights the current selection', () => {
    wrapper.find('[value="1"]').trigger('click')

    expect(wrapper.find('.selected').attributes('value')).toEqual('1')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('emits an `inputchanged` event when the selection changed', () => {
    wrapper.find('[value="1"]').trigger('click')

    expect(wrapper.emitted().inputchanged).toBeTruthy()
    expect(wrapper.emitted().inputchanged[0]).toEqual([1, false])
  })

  test('emits an `inputchanged` event with pencil mode if set', () => {
    wrapper.find('[value="1"]').trigger('click')
    wrapper.find('[value="pencil"]').trigger('click')

    expect(wrapper.emitted().inputchanged.length).toEqual(2)
    expect(wrapper.emitted().inputchanged[0]).toEqual([1, false])
    expect(wrapper.emitted().inputchanged[1]).toEqual([1, true])
  })

  test('highlights the pencil when in draft mode and another selection', () => {
    wrapper.find('[value="1"]').trigger('click')
    wrapper.find('[value="pencil"]').trigger('click')

    expect(wrapper.find('.selected:first-of-type').attributes('value')).toEqual('1')
    expect(wrapper.find('.selected:last-of-type').attributes('value')).toEqual('pencil')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
