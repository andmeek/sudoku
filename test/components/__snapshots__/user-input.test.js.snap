// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInput.vue disables completed numbers 1`] = `
<div class="user-input">
  <div class="numbers">
    <div><button value="1" class="disabled" disabled="disabled">1</button><button value="2" class="">2</button><button value="3" class="">3</button><button value="4" class="">4</button><button value="5" class="">5</button></div>
    <div><button value="6" class="">6</button><button value="7" class="">7</button><button value="8" class="">8</button><button value="9" class="">9</button> <button value="0" class="">
        <font-awesome-icon-stub icon="eraser" size="sm"></font-awesome-icon-stub>
      </button></div>
  </div>
  <div class="actions"><button value="pencil" class="">
      <font-awesome-icon-stub icon="pencil-alt" size="sm"></font-awesome-icon-stub>
    </button></div>
</div>
`;

exports[`UserInput.vue highlights the current selection 1`] = `
<div class="user-input">
  <div class="numbers">
    <div><button value="1" class="selected">1</button><button value="2" class="">2</button><button value="3" class="">3</button><button value="4" class="">4</button><button value="5" class="">5</button></div>
    <div><button value="6" class="">6</button><button value="7" class="">7</button><button value="8" class="">8</button><button value="9" class="">9</button> <button value="0" class="">
        <font-awesome-icon-stub icon="eraser" size="sm"></font-awesome-icon-stub>
      </button></div>
  </div>
  <div class="actions"><button value="pencil" class="">
      <font-awesome-icon-stub icon="pencil-alt" size="sm"></font-awesome-icon-stub>
    </button></div>
</div>
`;

exports[`UserInput.vue highlights the pencil when in draft mode and another selection 1`] = `
<div class="user-input">
  <div class="numbers">
    <div><button value="1" class="selected">1</button><button value="2" class="">2</button><button value="3" class="">3</button><button value="4" class="">4</button><button value="5" class="">5</button></div>
    <div><button value="6" class="">6</button><button value="7" class="">7</button><button value="8" class="">8</button><button value="9" class="">9</button> <button value="0" class="">
        <font-awesome-icon-stub icon="eraser" size="sm"></font-awesome-icon-stub>
      </button></div>
  </div>
  <div class="actions"><button value="pencil" class="selected">
      <font-awesome-icon-stub icon="pencil-alt" size="sm"></font-awesome-icon-stub>
    </button></div>
</div>
`;

exports[`UserInput.vue matches snapshot 1`] = `
<div class="user-input">
  <div class="numbers">
    <div><button value="1" class="">1</button><button value="2" class="">2</button><button value="3" class="">3</button><button value="4" class="">4</button><button value="5" class="">5</button></div>
    <div><button value="6" class="">6</button><button value="7" class="">7</button><button value="8" class="">8</button><button value="9" class="">9</button> <button value="0" class="">
        <font-awesome-icon-stub icon="eraser" size="sm"></font-awesome-icon-stub>
      </button></div>
  </div>
  <div class="actions"><button value="pencil" class="">
      <font-awesome-icon-stub icon="pencil-alt" size="sm"></font-awesome-icon-stub>
    </button></div>
</div>
`;
